pricegraph/src/orderbook/scalar.rs

```diff
@@ -0,0 +1,159 @@
+pub struct Price(f64);
+
+impl Price {
+    /// Creates a new price from a `f64` value. Returns `None` if the price
+    /// value is not valid. Specifically, it must be in the range `(0, +∞)`.
+    pub fn new(value: f64) -> Option<Self> {
+        if is_strictly_positive_and_finite(value) {
+            Some(Price(value))
+        } else {
+            None
+        }
+    }
+
+    /// Creates a new price from a `f64` value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the value is not valid.
+    pub fn from_raw(value: f64) -> Self {
+        Self::new(value).expect("invalid price value")
+    }
+
+    /// Creates a new price from an exchange price fraction.
+    pub fn from_fraction(price: &PriceFraction) -> Option<Self> {
+        if price.numerator != 0 && price.denominator != 0 {
+            Some(Price(assert_strictly_positive_and_finite(
+                price.numerator as f64 / price.denominator as f64,
+            )))
+        } else {
+            None
+        }
+    }
+
+    /// Gets the value as a `f64`.
+    pub fn value(self) -> f64 {
+        self.0
+    }
+
+    /// Converts a price into an effective exchange rate with explicit fees.
+    pub fn exchange_rate(self) -> ExchangeRate {
+        ExchangeRate(assert_strictly_positive_and_finite(self.0 * FEE_FACTOR))
+    }
+}
+
+/// An effective exchange rate that explicitly includes fees. As such, the
+/// actual exchange rate for a trade (the ratio of executed buy amount over
+/// executed sell amount) is greater than **or equal to** the limit exchange
+/// rate for an order.
+///
+/// Exchange rates are guaranteed to be a strictly positive finite real numbers.
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
+pub struct ExchangeRate(f64);
+
+impl ExchangeRate {
+    /// The 1:1 exchange rate.
+    pub const IDENTITY: ExchangeRate = ExchangeRate(1.0);
+
+    /// Gets the value as a `f64`.
+    pub fn value(self) -> f64 {
+        self.0
+    }
+
+    /// Converts an exchange rate into a price with implicit fees.
+    pub fn price(self) -> Price {
+        Price(assert_strictly_positive_and_finite(self.0 / FEE_FACTOR))
+    }
+
+    /// Computes the inverse exchange rate.
+    pub fn inverse(self) -> Self {
+        ExchangeRate(assert_strictly_positive_and_finite(1.0 / self.0))
+    }
+
+    /// Computes the edge weight for an exchange rate for the orderbook
+    /// projection graph.
+    ///
+    /// This is the base-2 logarithm of the exchange rate. This eanbles path
+    /// weights to be computed using addition instead of multiplication.
+    pub fn weight(self) -> f64 {
+        self.0.log2()
+    }
+}
+
+// NOTE: We can implement `Eq` for `ExchangeRate` since its value is garanteed
+// to not be `NaN`.
+impl Eq for ExchangeRate {}
+
+impl Ord for ExchangeRate {
+    fn cmp(&self, rhs: &Self) -> cmp::Ordering {
+        self.partial_cmp(rhs).expect("exchange rate cannot be NaN")
+    }
+}
+
+impl PartialEq<f64> for ExchangeRate {
```

Comment:
cool, could even be implemented for Limit Price if needed.

Reply:
For sure!